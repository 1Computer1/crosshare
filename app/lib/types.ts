import * as t from "io-ts";
import type { WordDBT } from './WordDB';

import { DBPuzzleT, CommentWithRepliesV } from '../lib/dbtypes';

export type Optionalize<T extends K, K> = Omit<T, keyof K>;
export type PartialBy<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;

export const BLOCK = ".";

export enum Direction {
  Across,
  Down
}

export interface WorkerMessage {
  type: string,
}
export interface AutofillResultMessage extends WorkerMessage {
  type: 'autofill-result',
  input: string[],
  result: string[]
}
export function isAutofillResultMessage(msg: WorkerMessage): msg is AutofillResultMessage {
  return msg.type === 'autofill-result';
}
export interface AutofillCompleteMessage extends WorkerMessage {
  type: 'autofill-complete',
}
export function isAutofillCompleteMessage(msg: WorkerMessage): msg is AutofillCompleteMessage {
  return msg.type === 'autofill-complete';
}
export interface LoadDBMessage extends WorkerMessage {
  type: 'loaddb',
  db: WordDBT,
}
export function isLoadDBMessage(msg: WorkerMessage): msg is LoadDBMessage {
  return msg.type === 'loaddb';
}
export interface AutofillMessage extends WorkerMessage {
  type: 'autofill',
  grid: string[],
  width: number,
  height: number
}
export function isAutofillMessage(msg: WorkerMessage): msg is AutofillMessage {
  return msg.type === 'autofill';
}

export interface Position {
  row: number,
  col: number,
}

export interface PosAndDir extends Position {
  dir: Direction,
}

const PuzzleJsonMandatoryV = t.type({
  title: t.string,
  size: t.type({
    rows: t.number,
    cols: t.number
  }),
  clues: t.type({
    across: t.array(t.string),
    down: t.array(t.string)
  }),
  grid: t.array(t.string)
});
const PuzzleJsonOptionalV = t.partial({
  highlighted: t.array(t.number),
  highlight: t.keyof({ circle: null, shade: null }),
});
export const PuzzleJsonV = t.intersection([PuzzleJsonMandatoryV, PuzzleJsonOptionalV]);
export type PuzzleJson = t.TypeOf<typeof PuzzleJsonV>;

const ClueV = t.type({
  num: t.number,
  dir: t.union([t.literal(0), t.literal(1)]),
  clue: t.string,
});
export type ClueT = t.TypeOf<typeof ClueV>;

export const PuzzleV = t.type({
  authorId: t.string,
  category: t.union([t.string, t.null]),
  authorName: t.string,
  moderated: t.boolean,
  publishTime: t.union([t.number, t.null]),
  title: t.string,
  size: t.type({
    rows: t.number,
    cols: t.number
  }),
  clues: t.array(ClueV),
  grid: t.array(t.string),
  highlighted: t.array(t.number),
  highlight: t.keyof({ circle: null, shade: null }),
  comments: t.array(CommentWithRepliesV),
});

export type PuzzleT = t.TypeOf<typeof PuzzleV>;
export type PuzzleResult = PuzzleT & { id: string };

export function puzzleTitle(puzzle: PuzzleT) {
  let title = puzzle.title;
  if (puzzle.category === 'dailymini' && puzzle.publishTime) {
    const d = new Date(puzzle.publishTime);
    const ds = (d.getUTCMonth() + 1) + "/" + d.getUTCDate() + "/" + d.getUTCFullYear()
    title = "Daily Mini for " + ds;
  }
  return title;
}

export function puzzleFromDB(dbPuzzle: DBPuzzleT): PuzzleT {
  let clues: Array<ClueT> = [];
  for (let i = 0; i < dbPuzzle.ac.length; i += 1) {
    clues.push({ dir: Direction.Across, clue: dbPuzzle.ac[i], num: dbPuzzle.an[i] });
  }
  for (let i = 0; i < dbPuzzle.dc.length; i += 1) {
    clues.push({ dir: Direction.Down, clue: dbPuzzle.dc[i], num: dbPuzzle.dn[i] });
  }
  return {
    authorId: dbPuzzle.a,
    category: dbPuzzle.c,
    authorName: dbPuzzle.n,
    moderated: dbPuzzle.m,
    publishTime: dbPuzzle.p ? dbPuzzle.p.toMillis() : null,
    title: dbPuzzle.t,
    size: {
      rows: dbPuzzle.h,
      cols: dbPuzzle.w
    },
    clues: clues,
    grid: dbPuzzle.g,
    highlighted: dbPuzzle.hs || [],
    highlight: dbPuzzle.s ? "shade" : "circle",
    comments: dbPuzzle.cs || []
  };
}
