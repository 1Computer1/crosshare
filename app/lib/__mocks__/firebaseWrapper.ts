const firebaseTesting = jest.requireActual('@firebase/testing');
import type firebaseAdminType from 'firebase-admin';

export let App: firebase.app.App;
export let AdminApp: firebaseAdminType.app.App;

export const setApp = (app: firebase.app.App) => { App = app; };
export const setAdminApp = (app: firebaseAdminType.app.App) => { AdminApp = app; };

export const AuthProvider = null;

export const DeleteSentinal = firebaseTesting.firestore.FieldValue.delete();

export const ServerTimestamp = firebaseTesting.firestore.FieldValue.serverTimestamp();

export const TimestampClass = firebaseTesting.firestore.Timestamp;
export const AdminTimestamp = firebaseTesting.firestore.Timestamp;

let anonApp: firebase.app.App;
let anonUser: firebase.User;
export const setUpForSignInAnonymously = (app: firebase.app.App, user: firebase.User) => {
  anonApp = app;
  anonUser = user;
};

export const signInAnonymously = async () => {
  setApp(anonApp);
  return anonUser;
};
