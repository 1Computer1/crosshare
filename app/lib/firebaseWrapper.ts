declare var firebase: typeof import('firebase');

// On the server we need to require firebase - on client it's done w/ script tags
if (typeof window === "undefined") {
  globalThis.firebase = require("firebase");
  require("firebase/firestore");
}

const firebaseConfig = {
  apiKey: "AIzaSyBrmmBf91peVT5T_Z7N3z9oizsPH5u2pUc",
  authDomain: "auth.crosshare.org",
  databaseURL: "https://mdcrosshare.firebaseio.com",
  projectId: "mdcrosshare",
  storageBucket: "mdcrosshare.appspot.com",
  messagingSenderId: "603173482014",
  appId: "1:603173482014:web:98d7d820731b7c5eaa080f",
  measurementId: "G-LTLN7Z4XBS"
};

// Initialize Firebase
export let App: firebase.app.App;
if (firebase.apps.length) {
  App = firebase.apps[0];
} else {
  App = firebase.initializeApp(firebaseConfig);
  if (typeof window !== "undefined") {
    if (process.env.NODE_ENV === 'production') {
      App.performance();
    }
  }
  if (process.env.NEXT_PUBLIC_USE_FIREBASE_EMULATOR) {
    var db = App.firestore();
    db.settings({
      host: "localhost:8080",
      ssl: false
    });
  }
}

export const setApp = (app: firebase.app.App) => { App = app };

export const AuthProvider = new firebase.auth.GoogleAuthProvider();

export const DeleteSentinal = firebase.firestore.FieldValue.delete();

export const TimestampClass = firebase.firestore.Timestamp;
export type TimestampType = firebase.firestore.Timestamp;
