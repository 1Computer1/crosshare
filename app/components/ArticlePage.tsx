import Head from 'next/head';
import { ArticleT } from '../lib/article';
import { ArticlePageProps } from '../lib/serverOnly';
import { HUGE_AND_UP, MAX_WIDTH } from '../lib/style';
import { ErrorPage } from './ErrorPage';
import { Link } from './Link';
import { Markdown } from './Markdown';
import { DefaultTopBar } from './TopBar';

export function ArticlePage(props: ArticlePageProps) {
  if ('error' in props) {
    return (
      <ErrorPage title="Article Not Found">
        <p>
          We&apos;re sorry, we couldn&apos;t find the article you requested.
        </p>
        <p>{props.error}</p>
        <p>
          Try the <Link href="/">homepage</Link>.
        </p>
      </ErrorPage>
    );
  }
  return <Article key={props.s} {...props} />;
}

function Article(props: ArticleT) {
  return (
    <>
      <Head>
        <title>{props.t}</title>
        <meta key="og:title" property="og:title" content={props.t} />
        <link
          rel="canonical"
          href={'https://crosshare.org/article/' + props.s}
        />
      </Head>
      <DefaultTopBar />
      <Markdown
        text={props.c}
        css={{
          margin: '2em 1em',
          [HUGE_AND_UP]: {
            maxWidth: MAX_WIDTH,
            margin: '2em auto',
          },
        }}
      />
    </>
  );
}
