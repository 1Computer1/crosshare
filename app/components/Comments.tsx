import { useState, useEffect, ReactNode, FormEvent } from 'react';
import * as t from 'io-ts';
import { isRight } from 'fp-ts/lib/Either';
import { PathReporter } from 'io-ts/lib/PathReporter';

import { PartialBy, Comment } from '../lib/types';
import { Identicon } from './Icons';
import { timeString } from '../lib/utils';
import { Emoji } from './Emoji';
import { buttonAsLink } from '../lib/style';
import { DisplayNameForm, getDisplayName } from './DisplayNameForm';
import { App, TimestampClass } from '../lib/firebaseWrapper';
import { CommentForModerationT, CommentForModerationWithIdV, CommentForModerationWithIdT } from '../lib/dbtypes';


const COMMENT_LENGTH_LIMIT = 140;

type LocalComment = Omit<Comment, 'id' | 'replies'>;
type CommentWithPossibleLocalReplies = Omit<Comment, 'replies'> & {
  replies?: Array<CommentOrLocalComment>
}
type CommentOrLocalComment = CommentWithPossibleLocalReplies | LocalComment

const SpoilerText = ({ text }: CommentTextProps) => {
  return <span css={{
    backgroundColor: 'var(--text)',
    '&:hover': {
      backgroundColor: 'var(--bg)',
    }
  }}>{text}</span>;
};

interface CommentTextProps {
  text: string
}
export const CommentText = ({ text }: CommentTextProps) => {
  const pieces: Array<ReactNode> = [];
  const tagFinder = /(?<item>(?<tag_begin>>!)(?<content>.+?)(?<tag_end>!<))/gm;
  let lastMatchedPosition = 0;
  let i = 0;
  function breaker(match: string, _item: string, _tag_begin: string, content: string, _tag_end: string, offset: number, string: string) {
    if (lastMatchedPosition < offset) {
      pieces.push(<span key={i++}>{string.substring(lastMatchedPosition, offset)}</span>);
    }
    pieces.push(<SpoilerText key={i++} text={content} />);
    lastMatchedPosition = offset + match.length;
    return match;
  }
  text.replace(tagFinder, breaker);
  pieces.push(<span key={i++}>{text.substring(lastMatchedPosition)}</span>);
  return <div>{pieces}</div>;
};

interface CommentProps {
  puzzleAuthorId: string,
  comment: CommentOrLocalComment,
  children?: ReactNode
}
const CommentView = (props: CommentProps) => {
  return (
    <div css={{
      marginTop: '1em',
    }}>
      <div><CommentFlair displayName={props.comment.authorDisplayName} userId={props.comment.authorId} puzzleAuthorId={props.puzzleAuthorId} solveTime={props.comment.authorSolveTime} didCheat={props.comment.authorCheated} /></div>
      <CommentText text={props.comment.commentText} />
      {props.children}
    </div>
  );
};

const CommentWithReplies = (props: PartialBy<CommentFormProps, 'user'> & { comment: CommentOrLocalComment }) => {
  const [showingForm, setShowingForm] = useState(false);
  const commentId = isComment(props.comment) ? props.comment.id : null;
  const replies = isComment(props.comment) ? props.comment.replies : undefined;
  return (
    <CommentView puzzleAuthorId={props.puzzleAuthorId} comment={props.comment}>
      {(!props.user || props.user.isAnonymous || !commentId) ?
        ''
        :
        (showingForm ?
          <div css={{ marginLeft: '2em' }}>
            <CommentForm {...props} onCancel={() => setShowingForm(false)} replyToId={commentId} user={props.user} />
          </div>
          :
          <div><button css={buttonAsLink} onClick={() => setShowingForm(true)}>Reply</button></div>
        )
      }
      {replies ?
        <ul css={{
          listStyleType: 'none',
          margin: '0 0 0 2em',
          padding: 0,
        }}>
          {replies.map((a, i) => <li key={i}><CommentWithReplies {... { ...props, comment: a }} /></li>)}
        </ul>
        :
        ''
      }
    </CommentView>
  );
};

function commentsKey(puzzleId: string) {
  return 'comments/' + puzzleId;
}

function commentsFromStorage(puzzleId: string): Array<CommentForModerationWithIdT> {
  const inSession = localStorage.getItem(commentsKey(puzzleId));
  if (inSession) {
    const res = t.array(CommentForModerationWithIdV).decode(JSON.parse(inSession));
    if (isRight(res)) {
      return res.right;
    } else {
      console.error('Couldn\'t parse object in local storage');
      console.error(PathReporter.report(res).join(','));
    }
  }
  return [];
}

interface CommentFlairProps {
  solveTime: number,
  didCheat: boolean,
  displayName: string,
  userId: string,
  puzzleAuthorId: string,
}
const CommentFlair = (props: CommentFlairProps) => {
  return (
    <>
      <span css={{ verticalAlign: 'text-bottom' }}><Identicon id={props.userId} /></span>
      <i> {props.displayName} </i>
      {(props.userId === props.puzzleAuthorId) ?
        <span css={{
          fontSize: '0.75em',
          backgroundColor: 'var(--primary)',
          color: 'white',
          borderRadius: 5,
          padding: '0.1em 0.2em',
        }}>constructor</span>
        :
        <>
          {props.didCheat ? <Emoji title='Used helpers' symbol='😏' /> : <Emoji title='Solved without helpers' symbol='🤓' />}
          < span css={{
            fontSize: '0.75em',
            backgroundColor: 'var(--caption)',
            color: 'white',
            borderRadius: 5,
            padding: '0.1em 0.2em',
          }}>{timeString(props.solveTime, false)}</span>
        </>
      }
    </ >
  );
};

interface CommentFormProps {
  displayName: string,
  setDisplayName: (name: string) => void,
  puzzleAuthorId: string,
  user: firebase.User,
  solveTime: number,
  didCheat: boolean,
  puzzleId: string,
  replyToId?: string,
}

const CommentForm = ({ onCancel, ...props }: CommentFormProps & { onCancel?: () => void }) => {
  const [commentText, setCommentText] = useState('');
  const [editingDisplayName, setEditingDisplayName] = useState(false);
  const [submittedComment, setSubmittedComment] = useState<LocalComment | null>(null);

  if (props.user === null) {
    throw new Error('displaying comment form w/ no user');
  }

  function sanitize(input: string) {
    return input.substring(0, COMMENT_LENGTH_LIMIT);
  }

  function submitComment(event: FormEvent) {
    event.preventDefault();
    const textToSubmit = commentText.trim();
    if (!textToSubmit) {
      return;
    }
    const comment: CommentForModerationT = {
      c: textToSubmit,
      a: props.user.uid,
      n: props.displayName,
      t: props.solveTime,
      ch: props.didCheat,
      p: TimestampClass.now(),
      pid: props.puzzleId,
      rt: props.replyToId !== undefined ? props.replyToId : null
    };
    console.log('Submitting comment', comment);
    const db = App.firestore();
    // Add to moderation queue for long term
    db.collection('cfm').add(comment).then((ref) => {
      console.log('Uploaded', ref.id);

      // Replace this form w/ the comment for the short term
      setSubmittedComment({
        commentText: comment.c,
        authorId: comment.a,
        authorDisplayName: comment.n,
        authorSolveTime: comment.t,
        authorCheated: comment.ch,
        publishTime: comment.p.toMillis(),
      });
      // Add the comment to localStorage for the medium term
      const forSession = commentsFromStorage(props.puzzleId);
      forSession.push({ i: ref.id, ...comment });
      localStorage.setItem(commentsKey(props.puzzleId), JSON.stringify(forSession));
    });
  }

  if (submittedComment) {
    return <CommentView puzzleAuthorId={props.puzzleAuthorId} comment={submittedComment} />;
  }

  return (
    <>
      <form onSubmit={submitComment}>
        <label css={{ width: '100%', margin: 0 }}>{(props.replyToId !== undefined ? 'Enter your reply ' : 'Leave a comment ') + '(please be nice!):'}
          <textarea css={{ width: '100%', display: 'block' }} value={commentText} onChange={e => setCommentText(sanitize(e.target.value))} />
        </label>
        <div css={{
          textAlign: 'right',
          color: (COMMENT_LENGTH_LIMIT - commentText.length) > 10 ? 'var(--default-text)' : 'var(--error)',
        }}>{commentText.length}/{COMMENT_LENGTH_LIMIT}</div>
        {
          editingDisplayName ?
            ''
            :
            <>
              <input css={{ marginRight: '0.5em', }} type="submit" disabled={commentText.length === 0} value="Save" />
              {onCancel ?
                <button type="button" css={{ marginRight: '0.5em' }} onClick={onCancel}>Cancel</button>
                : ''}
              commenting as <CommentFlair displayName={props.displayName} userId={props.user.uid} puzzleAuthorId={props.puzzleAuthorId} solveTime={props.solveTime} didCheat={props.didCheat} /> (
              <button css={buttonAsLink} onClick={() => setEditingDisplayName(true)}>change name</button>)
            </>
        }
      </form>
      {editingDisplayName ?
        <>
          <DisplayNameForm
            user={props.user}
            onChange={(s) => { props.setDisplayName(s); setEditingDisplayName(false); }}
            onCancel={() => setEditingDisplayName(false)}
          />
        </>
        :
        ''
      }
    </>
  );
};

interface CommentsProps {
  user?: firebase.User,
  solveTime: number,
  didCheat: boolean,
  puzzleId: string,
  puzzleAuthorId: string,
  comments: Array<Comment>
}

function isComment(comment: CommentOrLocalComment): comment is CommentWithPossibleLocalReplies {
  return 'id' in comment;
}

function findCommentById(comments: Array<CommentOrLocalComment>, id: string): CommentWithPossibleLocalReplies | null {
  for (const comment of comments) {
    if (!isComment(comment)) {
      continue;
    }
    if (comment.id === id) {
      return comment;
    }
    if (comment.replies !== undefined) {
      const res = findCommentById(comment.replies, id);
      if (res !== null) {
        return res;
      }
    }
  }
  return null;
}

export const Comments = ({ comments, ...props }: CommentsProps): JSX.Element => {
  const [toShow, setToShow] = useState<Array<CommentOrLocalComment>>(comments);
  const [displayName, setDisplayName] = useState(getDisplayName(props.user));

  useEffect(() => {
    const rebuiltComments: Array<CommentOrLocalComment> = comments;
    const unmoderatedComments = commentsFromStorage(props.puzzleId);
    const toKeepInStorage: Array<CommentForModerationWithIdT> = [];
    for (const c of unmoderatedComments) {
      const moderatedVersion = findCommentById(rebuiltComments, c.i);
      // The comment we saved in LS has already made it through moderation
      // so we don't need to merge it - by not adding to `toKeepInStorage` this
      // will also remove from LS
      if (moderatedVersion !== null) {
        console.log('found existing version, skipping one from LS');
        continue;
      }
      toKeepInStorage.push(c);
      const localComment: LocalComment = {
        commentText: c.c,
        authorId: c.a,
        authorDisplayName: c.n,
        authorSolveTime: c.t,
        authorCheated: c.ch,
        publishTime: c.p.toMillis(),
      };
      if (c.rt === null) {
        rebuiltComments.push(localComment);
      } else {
        const parent = findCommentById(rebuiltComments, c.rt);
        if (parent === null) {
          /* TODO figure out better behaivior here?
           * One possibility is that we saw the comment originally when
           * loading the puzzle via client side link but are now loading via
           * page refresh. If the content is cached at a different time the
           * parent comment might not be here yet. */
          continue;
        }
        if (parent.replies) {
          parent.replies.push(localComment);
        } else {
          parent.replies = [localComment];
        }
      }
    }
    // This means one or more have made it through moderation - update LS
    if (toKeepInStorage.length !== unmoderatedComments.length) {
      if (toKeepInStorage.length === 0) {
        localStorage.removeItem(commentsKey(props.puzzleId));
      } else {
        localStorage.setItem(commentsKey(props.puzzleId), JSON.stringify(toKeepInStorage));
      }
    }
    setToShow(rebuiltComments);
  }, [props.puzzleId, comments]);
  return (
    <div css={{ marginTop: '1em' }}>
      <h4 css={{ borderBottom: '1px solid var(--black)' }}>Comments</h4>
      {!props.user || props.user.isAnonymous ?
        <div>Sign in with google (above) to leave a comment of your own</div>
        :
        <CommentForm {...props} displayName={displayName} setDisplayName={setDisplayName} user={props.user} />
      }
      <ul css={{
        listStyleType: 'none',
        margin: '2em 0 0 0',
        padding: 0,
      }}>
        {toShow.map((a, i) => <li key={i}><CommentWithReplies displayName={displayName} setDisplayName={setDisplayName} comment={a} {...props} /></li>)}
      </ul>
    </div>
  );
};
