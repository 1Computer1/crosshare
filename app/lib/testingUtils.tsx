import { ReactElement, ComponentType } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { matchers, createSerializer } from 'jest-emotion';

import { AuthContext } from '../components/AuthContext';

import type firebaseTypes from 'firebase';

// Add the custom matchers provided by 'jest-emotion'
expect.extend(matchers);
expect.addSnapshotSerializer(createSerializer());

export const anonymousUser = {
  uid: 'anonymous-user-id',
  isAnonymous: true
} as firebaseTypes.User;

const WithAllProviders: ComponentType = ({ children }) => {
  return (
    <AuthContext.Provider value={{ user: anonymousUser, isAdmin: false, loadingUser: false, error: undefined }}>
      {children}
    </AuthContext.Provider>
  )
}

function wrappedRender(
  ui: ReactElement,
  options?: RenderOptions
): RenderResult {

  return render(ui, {
    wrapper: WithAllProviders,
    ...options,
  });
}
// re-export everything
export * from '@testing-library/react';

// override render method
export { wrappedRender as render };
