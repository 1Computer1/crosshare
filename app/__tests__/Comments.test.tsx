import { Comments } from '../components/Comments';
import { render } from '../lib/testingUtils';
import { Comment } from '../lib/types';
import { setApp } from '../lib/firebaseWrapper';
import * as firebaseTesting from '@firebase/testing';

jest.mock('../lib/firebaseWrapper');

const testComment: Comment = {
  id: 'comment-id',
  commentText: 'my first comment',
  authorId: 'comment-author-id',
  authorDisplayName: 'Mike D',
  authorSolveTime: 55.4,
  authorCheated: false,
  publishTime: new Date().getTime(),
};

test('basic comment display', () => {
  setApp(
    firebaseTesting.initializeTestApp({
      projectId: 'test1',
    }) as firebase.app.App
  );
  const { getByText, container } = render(
    <Comments
      solveTime={10}
      didCheat={false}
      puzzleId="puzz"
      puzzleAuthorId="puzzAuthor"
      comments={[testComment]}
    />,
    {}
  );
  expect(getByText(/my first comment/i)).toBeVisible();
  expect(container.firstChild).toMatchSnapshot();
});

test('comment with username display', () => {
  setApp(
    firebaseTesting.initializeTestApp({
      projectId: 'test1',
    }) as firebase.app.App
  );
  const { getByText, container } = render(
    <Comments
      solveTime={10}
      didCheat={false}
      puzzleId="puzz"
      puzzleAuthorId="puzzAuthor"
      comments={[{ authorUsername: 'MikeD', ...testComment }]}
    />,
    {}
  );
  expect(getByText(/my first comment/i)).toBeVisible();
  expect(container.firstChild).toMatchSnapshot();
});

test('security rules should only allow commenting as onesself', async () => {
  const app = firebaseTesting.initializeTestApp({
    projectId: 'mdcrosshare',
    auth: {
      uid: 'mike',
      admin: false,
      firebase: {
        sign_in_provider: 'google.com',
      },
    },
  });

  await firebaseTesting.assertFails(
    app.firestore().collection('cfm').add({ c: 'comment text' })
  );
  await firebaseTesting.assertFails(
    app.firestore().collection('cfm').add({ c: 'comment text', a: 'jared' })
  );
  await firebaseTesting.assertSucceeds(
    app.firestore().collection('cfm').add({ c: 'comment text', a: 'mike' })
  );
  app.delete();
});

test('security rules should only allow commenting with username if it matches your account', async () => {
  const adminApp = firebaseTesting.initializeAdminApp({ projectId: 'mdcrosshare' }) as firebase.app.App;
  await adminApp.firestore().collection('cp').doc('miked').set({ u: 'mike' });
  await adminApp.firestore().collection('cp').doc('rando').set({ u: 'rando' });
  adminApp.delete();

  const app = firebaseTesting.initializeTestApp({
    projectId: 'mdcrosshare',
    auth: {
      uid: 'mike',
      admin: false,
      firebase: {
        sign_in_provider: 'google.com',
      },
    },
  });



  await firebaseTesting.assertSucceeds(
    app.firestore().collection('cfm').add({ c: 'comment text', a: 'mike' })
  );
  await firebaseTesting.assertSucceeds(
    app.firestore().collection('cfm').add({ c: 'comment text', a: 'mike', un: 'miked' })
  );
  await firebaseTesting.assertSucceeds(
    app.firestore().collection('cfm').add({ c: 'comment text', a: 'mike', un: 'MikeD' })
  );
  await firebaseTesting.assertFails(
    app.firestore().collection('cfm').add({ c: 'comment text', a: 'mike', un: 'rando' })
  );
  await firebaseTesting.assertFails(
    app.firestore().collection('cfm').add({ c: 'comment text', a: 'mike', un: 'totalblast' })
  );
  app.delete();
});

test('security rules should only allow commenting if non-anonymous', async () => {
  const app = firebaseTesting.initializeTestApp({
    projectId: 'mdcrosshare',
    auth: {
      uid: 'mike',
      admin: false,
      firebase: {
        sign_in_provider: 'anonymous',
      },
    },
  });

  await firebaseTesting.assertFails(
    app.firestore().collection('cfm').add({ c: 'comment text' })
  );
  await firebaseTesting.assertFails(
    app.firestore().collection('cfm').add({ c: 'comment text', a: 'jared' })
  );
  await firebaseTesting.assertFails(
    app.firestore().collection('cfm').add({ c: 'comment text', a: 'mike' })
  );
  app.delete();
});
