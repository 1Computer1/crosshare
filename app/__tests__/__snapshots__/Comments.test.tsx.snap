// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic comment display 1`] = `
.emotion-0 {
  margin-top: 1em;
}

.emotion-1 {
  border-bottom: 1px solid var(--black);
}

.emotion-2 {
  text-align: center;
}

.emotion-3 {
  list-style-type: none;
  margin: 2em 0 0 0;
  padding: 0;
}

.emotion-4 {
  margin-top: 1em;
}

.emotion-4 p:last-of-type {
  margin-bottom: 0;
}

.emotion-5 {
  vertical-align: text-bottom;
}

.emotion-6 {
  font-size: 0.75em;
  background-color: var(--caption);
  color: white;
  border-radius: 5px;
  padding: 0.1em 0.2em;
}

.emotion-7 {
  font-style: italic;
}

<div
  class="emotion-0"
>
  <h4
    class="emotion-1"
  >
    Comments
  </h4>
  <div
    class="emotion-2"
  >
    <p>
      Sign in with google to leave a comment of your own:
    </p>
    <input
      alt="Sign in with Google"
      height="46"
      src="/googlesignin.png"
      type="image"
      width="191"
    />
  </div>
  <ul
    class="emotion-3"
  >
    <li>
      <div
        class="emotion-4"
      >
        <div>
          <span
            class="emotion-5"
          >
            <svg
              fill="rgb(25,42,230)"
              height="1em"
              shape-rendering="crispEdges"
              viewBox="0 0 5 5"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="rgba(140,140,140,0.2)"
                height="100%"
                width="100%"
              />
              <rect
                height="1"
                width="1"
                x="2"
                y="0"
              />
              <rect
                height="1"
                width="1"
                x="2"
                y="1"
              />
              <rect
                height="1"
                width="1"
                x="2"
                y="3"
              />
              <rect
                height="1"
                width="1"
                x="1"
                y="0"
              />
              <rect
                height="1"
                width="1"
                x="3"
                y="0"
              />
              <rect
                height="1"
                width="1"
                x="1"
                y="1"
              />
              <rect
                height="1"
                width="1"
                x="3"
                y="1"
              />
              <rect
                height="1"
                width="1"
                x="1"
                y="3"
              />
              <rect
                height="1"
                width="1"
                x="3"
                y="3"
              />
              <rect
                height="1"
                width="1"
                x="0"
                y="0"
              />
              <rect
                height="1"
                width="1"
                x="4"
                y="0"
              />
              <rect
                height="1"
                width="1"
                x="0"
                y="2"
              />
              <rect
                height="1"
                width="1"
                x="4"
                y="2"
              />
              <rect
                height="1"
                width="1"
                x="0"
                y="4"
              />
              <rect
                height="1"
                width="1"
                x="4"
                y="4"
              />
            </svg>
          </span>
          <i>
             
            Mike D
             
          </i>
          <span
            aria-label="Solved without helpers"
            role="img"
            title="Solved without helpers"
          >
            ðŸ¤“
          </span>
          <span
            class="emotion-6"
          >
            55s
          </span>
          Â Â·Â 
          <span
            class="emotion-7"
            title="2020-08-01T07:00:00.000Z"
          >
            1 day ago
             
          </span>
        </div>
        <div>
          <div
            class="paragraph"
          >
            my first comment
          </div>
        </div>
        
        
      </div>
    </li>
  </ul>
</div>
`;

exports[`comment with username display 1`] = `
.emotion-0 {
  margin-top: 1em;
}

.emotion-1 {
  border-bottom: 1px solid var(--black);
}

.emotion-2 {
  text-align: center;
}

.emotion-3 {
  list-style-type: none;
  margin: 2em 0 0 0;
  padding: 0;
}

.emotion-4 {
  margin-top: 1em;
}

.emotion-4 p:last-of-type {
  margin-bottom: 0;
}

.emotion-5 {
  vertical-align: text-bottom;
}

.emotion-6 {
  font-size: 0.75em;
  background-color: var(--caption);
  color: white;
  border-radius: 5px;
  padding: 0.1em 0.2em;
}

.emotion-7 {
  font-style: italic;
}

<div
  class="emotion-0"
>
  <h4
    class="emotion-1"
  >
    Comments
  </h4>
  <div
    class="emotion-2"
  >
    <p>
      Sign in with google to leave a comment of your own:
    </p>
    <input
      alt="Sign in with Google"
      height="46"
      src="/googlesignin.png"
      type="image"
      width="191"
    />
  </div>
  <ul
    class="emotion-3"
  >
    <li>
      <div
        class="emotion-4"
      >
        <div>
          <span
            class="emotion-5"
          >
            <svg
              fill="rgb(25,42,230)"
              height="1em"
              shape-rendering="crispEdges"
              viewBox="0 0 5 5"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="rgba(140,140,140,0.2)"
                height="100%"
                width="100%"
              />
              <rect
                height="1"
                width="1"
                x="2"
                y="0"
              />
              <rect
                height="1"
                width="1"
                x="2"
                y="1"
              />
              <rect
                height="1"
                width="1"
                x="2"
                y="3"
              />
              <rect
                height="1"
                width="1"
                x="1"
                y="0"
              />
              <rect
                height="1"
                width="1"
                x="3"
                y="0"
              />
              <rect
                height="1"
                width="1"
                x="1"
                y="1"
              />
              <rect
                height="1"
                width="1"
                x="3"
                y="1"
              />
              <rect
                height="1"
                width="1"
                x="1"
                y="3"
              />
              <rect
                height="1"
                width="1"
                x="3"
                y="3"
              />
              <rect
                height="1"
                width="1"
                x="0"
                y="0"
              />
              <rect
                height="1"
                width="1"
                x="4"
                y="0"
              />
              <rect
                height="1"
                width="1"
                x="0"
                y="2"
              />
              <rect
                height="1"
                width="1"
                x="4"
                y="2"
              />
              <rect
                height="1"
                width="1"
                x="0"
                y="4"
              />
              <rect
                height="1"
                width="1"
                x="4"
                y="4"
              />
            </svg>
          </span>
          <i>
             
            <a
              href="/MikeD"
            >
              Mike D
            </a>
             
          </i>
          <span
            aria-label="Solved without helpers"
            role="img"
            title="Solved without helpers"
          >
            ðŸ¤“
          </span>
          <span
            class="emotion-6"
          >
            55s
          </span>
          Â Â·Â 
          <span
            class="emotion-7"
            title="2020-08-01T07:00:00.000Z"
          >
            1 day ago
             
          </span>
        </div>
        <div>
          <div
            class="paragraph"
          >
            my first comment
          </div>
        </div>
        
        
      </div>
    </li>
  </ul>
</div>
`;
